src/coreclr/vm/jitinterface.cpp

```diff
@@ -14473,10 +14472,12 @@ CORINFO_METHOD_HANDLE CEEJitInfo::getAsyncResumptionStub()
     }
 
 #ifndef TARGET_X86
-    if ((msig.GetCallingConvention() & CORINFO_CALLCONV_PARAMTYPE) != 0)
+    if (msig.HasGenericContextArg())
     {
-        _ASSERTE(!"Generic context currently unhandled");
-        // Will need JIT to store this in an agreed upon place.
+        // TODO: Is the VM ok with this being null temporarily while the JIT is
+        // restoring state? Should we instead store the generic context (and
+        // perhaps "this") in a pre-agreed upon place, and then pass it here?
+        pCode->EmitLDC(0);
         numArgs++;
```

Comment:
Any idea about this? I can imagine GC reporting might not be happy about seeing the null generic context in case we suspend before its value has been restored from the continuation inside the target method.

Reply:
I think there is also a GC hole with collectible assemblies we'll have to solve in some way: the continuation does not necessarily have anything keeping the loader allocator for the generic context alive. It seems like for shared generics we'll have to add the loader allocator's rooting object for the generic context arg into the array of GC objects in the continuation. Even for unshared code we have to do this when emitting code into a collectible assembly.

This seems like it will probably a helper call in the suspension case to get the loader allocator from the generic context/target method. The other alternative is that GC scanning could iterate all continuations and report their target methods/generic contexts, but that does not seem like a good solution (we really do not want to have to iterate all continuations outside the standard optimized GC paths).

Reply:
> Any idea about this? I can imagine GC reporting might not be happy about seeing the null generic context in case we suspend before its value has been restored from the continuation inside the target method.

I think having null temporarily should be ok. As long as continuation is still reachable, which it should be until we restore.

Reply:
>  The other alternative is that GC scanning could iterate all continuations and report their target methods/generic contexts, but that does not seem like a good solution (we really do not want to have to iterate all continuations outside the standard optimized GC paths).

Right. It would be very expensive in Gen0 collections. Gen0 must be able to scan only a small (young) subset of continuations.

Adding the exposed object to the continuation seems the right way. Even if we do not use it for restoring anything - just as a keep-alive reference.

Reply:
> It may hit some asserts and they may need to be tweaked

I think I will switch this over to load the generic context in the caller once we have the JIT reporting back the continuation layout to the EE. For now we can probably keep it like this.

I'm going to look into addressing the GC holes in a follow-up.

Reply:
Allowing the context arg to be NULL is not a big deal as long as we are safely preserving the exposed LoaderAllocator object somewhere. We would have to remove those asserts, but as you can see the code already handles the param being marked as NULL. In fact, if we can guarantee that the LoaderAllocator exposed object is kept live, the runtime has no need to require that the context as reported by the CrawlFrame ever be set. (Please don't do this, as it reduces the quality of the call stack, but a temporary time as NULL is not a meaningful problem).

Reply:
> In fact, if we can guarantee that the LoaderAllocator exposed object is kept live, the runtime has no need to require that the context as reported by the CrawlFrame ever be set.

I think the regular stackwalk needs to report the Loader from the context, as there may be no other reference to the exposed object. It seems hard to prove that for a live frame with unloadable context there is always something else that keeps the loader alive.

The suspended frames are not stackwalked though, thus we need to keep the exposed object reachable as long as continuation is reachable. Stashing a reference to the exposed object in the continuation should be sufficient, I think.

Reply:
Right, that was what I was thinking about. We need to hold onto the continuation until the context is set in the method frame (and the continuation needs to hold the exposed object), but otherwise, we don't need the thing to be non-null until general code execution begins.

Reply:
If that sounds ok to you then I will just keep it this way -- I've removed the assert in GC about the non-null generic context and the TODO comment here.